core/src/main/scala/chisel3/Annotation.scala

```diff
@@ -23,16 +28,42 @@ trait ChiselAnnotation {
   *
   *  Defines a conversion to corresponding FIRRTL Annotation(s)
   */
+@deprecated(
+  "Avoid custom annotations. If you must use annotations, new annotate.apply method that takes Data",
+  "Chisel 7.0"
+)
 trait ChiselMultiAnnotation {
   def toFirrtl: Seq[Annotation]
 }
 
+@nowarn("msg=Avoid custom annotations")
 object annotate {
-  def apply(anno: ChiselAnnotation): Unit = {
-    Builder.annotations += anno
-  }
-  def apply(annos: ChiselMultiAnnotation): Unit = {
-    Builder.newAnnotations += annos
+  @deprecated(
+    "Avoid custom annotations. If you must use annotations, new annotate.apply method that takes Data",
+    "Chisel 7.0"
+  )
+  def apply(anno: ChiselAnnotation): Unit = apply()(Seq(anno.toFirrtl))
+
+  @deprecated(
+    "Avoid custom annotations. If you must use annotations, new annotate.apply method that takes Data",
+    "Chisel 7.0"
+  )
+  def apply(annos: ChiselMultiAnnotation): Unit = apply()(annos.toFirrtl)
+
+  /** Create annotations.
+    *
+    * Avoid this API if possible.
+    *
+    * Anything being annotated must be passed as arguments so that Chisel can do safety checks.
+    * The caller is still responsible for calling .toTarget on those arguments in mkAnnos.
+    */
+  def apply(targets: InstanceId*)(mkAnnos: => Seq[Annotation]): Unit = {
+    targets.foreach {
+      case d: Data => requireIsAnnotatable(d, "Data marked with annotation")
+      case _ => ()
+    }
+    // TODO record views that need to be renamed
+    Builder.annotations += (() => mkAnnos)
   }
```

Comment:
Is there a way to do this which directly passes the `targets` to the annotation construction function directly?

For this to work, an `Annotation` (or a subclass of) would need to have a constructor that can build it from an `InstanceId` (or a `Data`).

I'm primarily worried about there being no check of the connection between the `targets` and the `mkAnnos` function as a user could put in incorrect targets or ignore them entirely to get the old, legacy behavior.

Can you think of a way to untangle this?

Reply:
Discussed offline--we could untangle this, but it's a lot of extra infrastructure around stuff that users aren't really supposed to be using anyway since user-defined annotations are not supported in CIRCT. This is slightly ugly but good enough.